### PR TITLE
Adding `bucket_as_host` flag to generate signed url

### PR DESCRIPTION
**Pull request by gr8Adakron**
The following PR will allow us to create signed url in the following nomenclature: `https://custom.domain.com/image.jpg?` which wasn't feasible before with this library. This allow us to be independent of amazon host. As in real time scenario when generating signed url for clients side application the IP whitelisting is a well known problem. Certainly this resolve the prior issue as one get better control over custom domain IPs. 
Further benefits with been custom domain as host - provision us to configure authentication layer on access of `presigned` url too.

Inspired from the following issues:
- https://github.com/aws/aws-sdk-js/issues/891
- https://github.com/ex-aws/ex_aws_s3/issues/154

### CODE REVIEW

**Comment by gr8Adakron** [comment]
@bernardd FYI

**Comment by bernardd** [comment]
Hi @gr8Adakron - the change itself looks good; just a couple of little requests:
* I think the `s3_bucket_endpoint` option might be better named something like `bucket_as_host` (unless AWS use the former nomenclature somewhere).
* Could you add the new option, whatever it ends up being called, to the `presigned_url_opts()` type spec
* Could you also update the docs for `presigned_url()` to reflect the new option (I just noticed the existing docs for `virtual_host` are slightly wrong, too, which is probably what caused the confusion over in #154).

**Comment by gr8Adakron** [comment]
@bernardd I choose this `s3_bucket_endpoint` argument as it was already been considered in other libraries. 
- I was thinking to add separate PR for documentation update.

**Comment by bernardd** [comment]
Also it would be great to have a test case to cover this use.

**Comment by gr8Adakron** [comment]
@bernardd 
- Renamed flag to `bucket_as_host` - as its easy to understand.
- Adding REAME.md for `bucket_as_host` flag
- Adding code level doc for `bucket_as_host` flag
- Fixing code level doc for `virtual_host` flag
- Adding test cases for `bucket_as_host` flag in coupled with `virtual_host` flag.

Please review.

**Comment by bernardd** [comment]
Thanks @gr8Adakron - that all looks good. One tiny little request. The docs say:

>   When option param `:bucket_as_host` is `true`, the bucket name will be used as the hostname, which will look like `<bucket>.<domain>.com` host.

but actually `bucket` is the full hostname, there's no additional `domain` or `.com` parts. Could you reword it to something like

> When option param `:bucket_as_host` is `true`, the bucket name will be used as the full hostname. In this case, `bucket` must be set to a full hostname, for example `mybucket.example.com`.

If you can make that tweak, I'll get this all merged. Thanks for all your work on it :)

**Comment by gr8Adakron** [comment]
@bernardd Updated for the mentioned tweak.

**Comment by bernardd** [comment]
Fantastic, thanks heaps for your work on this @gr8Adakron!
